### ISSUE TITLE
rpm without TTY gives wrong error message

### ISSUE THREAD

**Issue by sourcejedi** [rpm-software-management/rpm#355]
```
$ rpm -e mlocate
error: can't create transaction lock on /var/lib/rpm/.rpm.lock (Permission denied)
$ rpm -e mlocate </dev/null
error: can't create transaction lock on /var/lib/rpm/.rpm.lock (Inappropriate ioctl for device)

$ rpm -q rpm
rpm-4.13.0.2-1.fc26.x86_64
```
This is unfortunate for automation e.g. Ansible.

**Comment by n3npq** [comment]
Run as root.

**Comment by sourcejedi** [comment]
I don't understand.  What are you asking for?

This is a UI bug, triggered when a non-query `rpm` command is run without permissions _and_ without a tty on STDIN.

I don't see what more information I can provide.  Running `rpm` _with_ permissions doesn't print an error message; I checked and it doesn't matter whether I redirect STDIN or not.

If you're saying this is issue closed as some flavour of WONTFIX, you forgot to close it.

**Comment by n3npq** [comment]
I assumed you wished rpm --erase to function.

The important part of the error message is correct even if errno is incorrect when stdin is redirected.

FWIW, I can't think of any rpm mode that actually reads from stdin (passwords are read using getpass(3)).

Meanwhile errno is (likely) being reset by isatty(3) trying to detect whether to colorize output.

**Comment by andreas-schwab** [comment]
It's actually rpmlock_acquire not setting errno when it returns with a failure when asked to write-lock a read-only lock.

**Comment by sourcejedi** [comment]
Yeah, I just found it too.  In principle the closest errno would be EBADF though, which is not much more user-friendly :(.  So I'm not sure what the best response to this is.

>        EBADF  cmd  is  F_SETLK  or  F_SETLKW and the file descriptor open mode
>              doesn't match with the type of lock requested.

(same as `write()` on an fd which was not opened for writing).

**Comment by pmatilai** [comment]
In the meanwhile I manually merged this one with somewhat different commit message, the EROFS-specific test was dumb anyhow.

**Comment by pmatilai** [comment]
First part was already fixed in commit df053dd36461042e4d11a21a97e9be9058204658, second half now in commit b89eb35c6034d43a8048ac9d452a1921596071a7:

```
[pmatilai@sopuli rpm]$ ./rpm -e mlocate
error: can't create transaction lock on /var/lib/rpm/.rpm.lock (Permission denied)
[pmatilai@sopuli rpm]$ ./rpm -e mlocate < /dev/null
error: can't create transaction lock on /var/lib/rpm/.rpm.lock (Permission denied)
```
